feat: support refresh token rotation - alternative

strehle opened the pull request:
@bruce-ricard  @peterhaochen47 
Same logic done with PR https://github.com/cloudfoundry/uaa/pull/1969
But will less changes in large UaaTokenServices and more in RefreshTokenCreator

https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2044

strehle committed:
feat: support refresh token rotation

- switch: jwt.token.refresh.rotate
- default off / false
- Needed for OAuth 2.1 compliance

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokens

strehle committed:
add tests

strehle committed:
add unit test for refresh token rotation

strehle committed:
add junit test

strehle committed:
add revert junit test here

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokens

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokens

strehle committed:
Merge branches 'feat/rotateRefreshTokens' and 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokens

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokens

strehle committed:
feat: support refresh token rotation - alternative PR

see https://github.com/cloudfoundry/uaa/pull/1969

bruce-ricard commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java:
```diff
@@ -339,6 +338,9 @@ public OAuth2AccessToken refreshAccessToken(String refreshTokenValue, TokenReque
         );
 
+        if (isRevocable && refreshTokenCreator.isRefreshTokenRotate()) {
+            tokenProvisioning.delete((String) tokenValidation.getClaims().get(JTI), -1, IdentityZoneHolder.getCurrentZoneId());
```
I'm confused about the deletion here. I was expected to see a new refresh token be created next to the deletion of the old one in case the rotation was triggered.

strehle replied:
done

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/DeprecatedUaaTokenServicesTests.java:
```diff
@@ -1908,6 +1910,36 @@ public void refreshAccessToken_withAccessToken() {
     }
 
+    @Test
+    public void refreshRotation() {
```
I forgot the context around this `DeprecatedUaaTokenServicesTests.java` file, whose name says "deprecated". Why is it still around? Do we really want to add tests to it?

bruce-ricard commented on uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/OIDCLoginIT.java:
```diff
@@ -508,6 +517,18 @@ public void successfulLoginWithOIDC_and_SAML_Provider() throws Exception {
+            setRefreshTokenRotate(true);
+            refreshToken1 = getRefreshTokenResponse(serverRunning, refreshToken2);
+            refreshToken2 = getRefreshTokenResponse(serverRunning, refreshToken1);
+            assertNotEquals("New access token should be different to the old one.",
```
"different **from** the old one"

bruce-ricard commented on model/src/main/java/org/cloudfoundry/identity/uaa/zone/TokenPolicy.java:
```diff
@@ -116,6 +117,14 @@ public void setRefreshTokenUnique(boolean refreshTokenUnique) {
         this.refreshTokenUnique = refreshTokenUnique;
     }
 
+    public boolean isRefreshTokenRotate() {
```
The name of this getter is awkward to understand to me. What about something like `shouldRotateRefreshTokens()` ?

strehle replied:
TokenPolicy is used to read JSON data and here is<variable> is need, but I renamed the attribute from RefreshTokenCreator

bruce-ricard reviewed (changes requested):
Sorry for taking so long to revew!

strehle committed:
changes from review

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java:
```diff
@@ -724,6 +728,9 @@ CompositeToken persistRevocableToken(String tokenId,
             }
             tokenProvisioning.createIfNotExists(revocableRefreshToken, IdentityZoneHolder.get().getId());
+            if (tokenIdToBeDeleted != null) {
+                tokenProvisioning.delete(tokenIdToBeDeleted, -1, IdentityZoneHolder.getCurrentZoneId());
```
moved the delete direct after the new creation

strehle commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/DeprecatedUaaTokenServicesTests.java:
```diff
@@ -1908,6 +1910,37 @@ public void refreshAccessToken_withAccessToken() {
         tokenServices.refreshAccessToken(getOAuth2AccessToken().getValue(), getRefreshTokenRequest());
     }
 
+    //@Test
```
I will check in sonar. I think I placed the tests here because of test coverage

strehle committed:
add new test class for refresh rotation

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokensAlternative

strehle committed:
delete from deprecated class

strehle committed:
refactor test

strehle committed:
test

strehle committed:
move test

strehle committed:
reformatting

strehle committed:
remove from this class

UaaTokenServicesTests is in wrong package uaa instead of server,
therefore sonarcube does not cover the tests

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feat/rotateRefreshTokensAlternative

strehle committed:
fix

strehle committed:
fix
